pyquil/analytics.py

```diff
@@ -0,0 +1,310 @@
+                         using 1, 2, 3, ... qubits without indicating the range of indices for these
+                         qubits. Default is False.
+    :param to_include: (Optional) A list of properties to analyze with your program creator. Default is
+                       to include every parameter.
+    :param remove_zeros: (Optional) If set to True, any properties where the value is 0 for all inputs
+                         of the program are left out of the final result. Default is False.
+    :return: A tuple containing multiple analysis results:
+                0) The input list used to generate each program
+                1) A structure of arrays, with indices corresponding to the index of the input in input_list,
+                and keys representing the properties measured for those program compilations
+                2) A dictionary from properties to possible complexities for that property (wrt input)
+    '''
+
+    results = []
+    for val in input_list:
+        # Space efficient option to pass in sizes of a list
+        # rather than a list of increasing lists. This is helpful
+        # for items such as qubit ranges
+        if should_range:
+            val = range(val)
+        
+        start_time = time.time() / 1000.0
+        program = program_creator(val)
+        properties = get_qubit_num_gates(get_gate_types(program))
+        properties["Total Instructions"] = get_total_instruction_count(program)
+        properties["Total Qubits"] = get_qubit_count(program)
+        properties["Compile Time"] = time.time() / 1000.0 - start_time
+        results.append(properties)
+        
+    results = aos_to_soa(results)
+    
+    # Filters results if specific ones are requested
+    if to_include != None:
+        filtered = {}
+        for val in to_include:
+            if val in results:
+                filtered[val] = results[val]
+        results = filtered
+    
+    # Removes properties that are always zero, if requested
+    if remove_zeros:
+        for key in results.keys():
+            if sum(results[key]) == 0:
+                del results[key]
+    
+    return (input_list, results)
+
+def plot_analysis(x_points, analysis, title, x_axis, y_axis, bounds=None):
+    '''
+    Takes results from the analyze function and plots them using plotly
+    :param analysis: The analysis gathered by the analyze function
+    :param title: The title to display on the plotted graph
+    :param x_axis: A label for the x axis
+    :param y_axis: A label for the y axis
+    :param bounds: An optional dictionary of names to functions, which will
+                   be plotted against the given x points. This is useful
+                   for comparing your gate growth to the requested growth
+    :param x_points: A list of x points to use for your data (i.e. input
+                     for your programs, "n")
+    '''
+    runtime = {}
+    data = []
+    
+    r = lambda: random.randint(0,255)
```

Comment:
avoid single letter variable names as they break IDE tools.  even using `rr` would be better.  There's not need to save on chars and semantic names go a long long way